Avoid NPE while fetching offline sessions

mkrueger92 opened the pull request:
It might happen that a user has been removed on a federation provider side. But the offline session is still in the system. This patch adds a check to prevent NPE when fetching all offline sessions through the UI.

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

pedroigor reviewed (commented):
@mkrueger92 I'm wondering if the proper way to deal with this situation is to make sure sessions are removed when disabling/removing users though a federation provider.

I thought we were doing that already. 

@mposolda Any thoughts?

pedroigor commented:
I see. So perhaps we are only left with the changes here in the proposed but we could probably remove the session when this situation happens rather than keep it there. This should be especially important for offline sessions.

For me, the we can fix at least this part of the code and do others as a follow-up.

mkrueger92 commented:
@pedroigor Not sure if a background task is needed in that case. But that would mean that this task needs to iterate over all sessions and check if a user is still available for it. Is that a good idea?

pedroigor commented:
@mkrueger92 I meant to remove the session when `user == null` rather than keep it there until expires.

@mposolda Does it make sense?

mkrueger92 commented:
@pedroigor from my point of view it would not do any harm. But since the code I adjusted is called manually by user interaction, there is no guarantee that sessions will be remembered bed ever. But it could be a start to minimize the problem.

I suggest a staged approach. First fix the NPE, which is a real problem for me now and second think about the right approach to deal with stale sessions. What do you think?

mposolda reviewed (changes requested):
2 points:

1) Can we handle also `UserResource.getSessions()` same way and do the same for "online" sessions as well? As this problem is not specific only to offline sessions, but this NPE can happen for normal sessions as well. Likely just something in `UserResource.getSessions()` might do the trick?

```
return session.sessions().getUserSessionsStream(realm, user)
        .filter(userSession -> userSession().getUser() != null)
        .map(ModelToRepresentation::toRepresentation);
```


2) How about some testing for this? Maybe it's possible to have some automated test, which will do:
- Login user
- Delete user in the federation provider
- Call the sessions endpoint (or offline sessions endpoint) with admin client for example by `UserResource.getUserSessions()` and check the sessions is not returned

The most tricky part is likely deleting user in the federation provider. I am not sure if we have something like this in `UserStorageTest` (you can check), but if not, I am pretty sure this can be tested easily with LDAP. For example we have test like `LDAPProvidersIntegrationTest.testLDAPUserRefreshCache`, which removes user directly from LDAP and then tests something, so maybe that one can be used as an inspiration.

pedroigor commented:
@mposolda What about removing the session rather than keeping it there until it expires?

mposolda commented:
> @mposolda What about removing the session rather than keeping it there until it expires?

@pedroigor @mkrueger92 I don't have strong preference. Whatever you guys prefer to do.

Maybe I am rather more inclined to keep the sessions as infinispan should expire them later on automatically. Also AFAIK REST "Get" endpoints are supposed to be "idempotent" when possible? And finally, when user has big amount of sessions, the "GET" endpoint can block for longer time when removing sessions (however this would be rather bigger problem for realms and clients, which can have much bigger amount of sessions than users, who typically have just a few sessions). On the other hand, removing sessions has also some advantages. So if you both agree to remove, I am fine with it.

mkrueger92 commented:
@mposolda: I totally agree that normal sessions might be effected too. I have changed the code accordingly. I had a look at the test code for LDAP but was kind of overwhelmed by it. Without additional hints I am not able to come up with a useful test.
Along the way I fixed another NPE we have seen in our logs resulting from the same issue.

Please advice what to do next.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java:
```diff
@@ -514,8 +514,10 @@ public Stream<UserSessionRepresentation> getUserSessions(@QueryParam("first") In
+        return session.sessions()
+                .getUserSessionsStream(client.getRealm(), client, firstResult, maxResults)
+                .map(ModelToRepresentation::toRepresentation)
+                .filter(Objects::nonNull);
```
It should be safer to filter sessions in `InfinispanUserSessionProvider` and avoid changing its callers.

mkrueger92 replied:
I am not sure about this.
The consequence of your suggestion is to tightly couple the `UserSessionProvider` and the `UserProvider`.
It must be defined in the interface contract what happens to sessions without a valid user (what about deactivated users?). But then the function should be renamed to something like "_getValidUserSessionsStream()_".
Before changing the interface in such a way, all users of this interface must be looked at.

pedroigor replied:
Looking at the places where the method is used, they only want valid sessions.

If we need to for whatever reason to handle the cases where sessions are not bound to users, we can provide a separate method or specific handing within the session provider itself.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java:
```diff
@@ -560,8 +562,10 @@ public Stream<UserSessionRepresentation> getOfflineUserSessions(@QueryParam("fir
-                .map(this::toUserSessionRepresentation);
+        return session.sessions()
+                .getOfflineUserSessionsStream(client.getRealm(), client, firstResult, maxResults)
+                .map(this::toUserSessionRepresentation)
```
It should be safer to filter sessions in `InfinispanUserSessionProvider` and avoid changing its callers.

mkrueger92 replied:
Same comment as above https://github.com/keycloak/keycloak/pull/17577#discussion_r1162283337

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java:
```diff
@@ -790,6 +794,9 @@ private void updateAuthorizationSettings(ClientRepresentation rep) {
      */
     private UserSessionRepresentation toUserSessionRepresentation(final UserSessionModel userSession) {
         UserSessionRepresentation rep = ModelToRepresentation.toRepresentation(userSession);
+        if (rep == null) {
```
If you are filtering at the source (`InfinispanUserSessionProvider`) you shouldn't worry about sessions not bound with a user.

mkrueger92 replied:
I don't think that is a good idea. The interface of `UserLookupProvider.getUserById(RealmModel realm, String id) `states that `null` is a valid return value. That means, every call to this function might produce `null` as result. Unless `InfinispanUserSessionProvider` returns the UserModel somehow in its data structure a null check of the caller is required. Especially in such constructs as `session.users().getUserById(getRealm(), entity.getUserId());`. This is used quite often. There you never know how a the interface `UserLookupProvider` is implemented and which constrains are attached to the component.

pedroigor replied:
As mentioned before, the places where the session is being used expect valid sessions, bound to a user.

If that is true. There is no reason to duplicate code (filtering by user not null) everywhere.

@mposolda Could confirm if my assumption is correct ....

mkrueger92 replied:
And as I mentioned before, `session.users().getUserById(realm, entity.getUser())` may return `null` any time. As long as `org.keycloak.models.sessions.infinispan.UserSessionAdapter.getUser()` still calls `session.users().getUserById(realm, entity.getUser())` there is a chance that `getUser()` returns `null`. And this must be checked, regardless if it has been checked before. This function is called at `ModelToRepresentation.toRepresentation(UserSessionModel session)`.

Also, in case the `InfinispanUserSessionProvider` checks the user for each element in the stream, this means that every time a lookup of the user is made (using a federated provider; LDAP, or a custom one, like in our case). And then again when the result is used, because `getUser()` makes the call again!

This results in a lot of round-trips! Even from a performance point of view, I see this a bad design.

In my opinion the first thing to be done is to define the expectations of the interfaces. Is there a design document somewhere?

Please recall, I started this PR to fix some NPEs. This is still my goal. I do not want to redesign this whole thing. But with 150k+ concurrent sessions in our system I do not see that our infrastructure is pretty happy doubling the number of requests just because the check for a valid user is done deep down the stack. (Again, it needs to be done at every call of `session.users().getUserById(...)`!)

mkrueger92 replied:
Nevertheless and regardless of my concerns, if you insist I can rewrite this PR to be closer to what I assume is your preferred solution. I need the NPEs fixed. Please let me know your decision.

pedroigor replied:
I had the wrong assumption that `UserSessionAdapter` (infinispan)  was leveraging the user cache. I see the problem when resolving the user from the session.

Looks like there are other things to evaluate here but unrelated to this PR. For instance, currently we are still making multiple calls to the underlying storage if `UserSessionAdapter.getUser()` is called multiple times during the request lifecycle.

Changes LGTM but we need some tests. Are you familiar with our test suite or do you want some help?

mposolda replied:
I see both points.

I agree that currently when `UserSessionProvider` returns some sessions, it doesn't check if user connected with the userSession exists. So we assume that it is responsibility of the caller to check if `userSession.getUser()` is null (and `userSession.getUser()` calls lookup of user in `UserProvider`).

If we change this behaviour, that `UserSessionProvider` returns only sessions, where user was proved to exists. This has big advantage, that we won't need to check `if (userSession.getUser() == null)` on various places inside codebase, which would be very great. Basically all sessions returned by userSessionProvider would be guaranteed to be valid sessions with valid user.

On the other hand, I am not sure about performance side-effect of this? As `UserSessionProvider` will need to verify existence of the user before the userSession is returned from `UserSessionProvider`. However performance side-effects may not be that big because we use Stream API?

I think it may be good to involve store team for this discussion.

@hmlnarik any opinion?

Should we just go with the changes in this PR (so verify nulls on the caller) and then consider whether we want to improve implementation of userSessionProvider itself? In that case, I think we should do it for both old and new store.

mkrueger92 replied:
In case the user is checked in the `UserSessionProvider` then the session object should be changed to already include the user and do not look it up again.

The performance side-effect comes into play when (like in our case) the user federation provider is configured with a NOCACHE policy. Then every call to `session.users().getUserById(...)` ends up at the provider and produces network traffic and so on. Therefore the user lookup should be reduced to a minimum. And the Stream API does no reduce the amount of user lookups at all.

mkrueger92 replied:
> Changes LGTM but we need some tests. Are you familiar with our test suite or do you want some help?

Unfortunately, I am not familiar with the test suite. I will have a deeper look next week.

mhajas replied:
Hello all, this is a tough decision to make. 

We had a call with @hmlnarik around this and we believe this needs to be considered case by case. 

In general, moving such integrity checks to the storage makes sense. It results in cleaner logical layer code as well as performance gain as we stop the processing for invalid entities sooner in the process chain. However, the latter doesn't need to be true in all cases. For example, if we will check all parent groups when obtaining a group from storage we can end up with a huge performance impact. 

For this specific use-case, I would lean towards the suggestion from @pedroigor and @mposolda. Sessions for not existing users are invalid and in a perfect world, they would be invisible to Keycloak (for example, if everything would be stored in one database, the sessions would be removed as part of some cascade removal). In cases such as this, where the user was removed from an external system we need to do our best to find this out as soon as possible and therefore I would add this check to the storage. Similarly, we are doing this for client sessions:
https://github.com/keycloak/keycloak/blob/4f8d67c9fcbc1a50976128584369d9a1a9077ca2/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/UserSessionAdapter.java#L110-L126

Note: to avoid loading the user again and again from the LDAP, we can change the `getUser` getter to something like:
```java
public UserModel getUser() {
    if (this.user == null) {
        this.user = session.users().getUserById(realm, entity.getUser());
    }

    return this.user;
}

```
CC: @hmlnarik

mkrueger92 replied:
@mhajas + @hmlnarik: I will change the PR and come up with a solution closer to your suggestion. Stay tuned.

mhajas replied:
Thank you @mkrueger92 !

pedroigor reviewed (changes requested):
@mkrueger92 Thanks.

We also need to make changes to the new store at `org.keycloak.models.map.userSession.MapUserSessionProvider`. @keycloak/store-maintainers, does it make sense?

Commit:
Optimize user lookup

* This commit adds some caching. Especially when federation is
  used, every call to getUser() might be expensive.

* Especially for offline sessions, it might happen that
  users are delete from federated backend without Keycloak
  knowing about it. This commit adds a check for that.

Commit:
Make sure to not return null elements in stream

mkrueger92 commented:
@mposolda @pedroigor @mhajas @abbra: I have reworked my approach. Please recheck the result.

Now, the user lookup is done early and cached. Please advice if I missed something.

Thx.

mhajas commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -739,7 +740,17 @@ protected void removeUserSession(UserSessionEntity sessionEntity, boolean offlin
+
+        UserModel user =  session.users().getUserById(realm, entity.getUser());
+        if (user == null) {
+            return null;
```
Should we remove the user session before we return `null`?
```suggestion
            removeUserSession(entity, offline);
            return null;
```

mkrueger92 replied:
I think the session removal should not be done.
I have the following reasons:
- The session will time out sooner or later anyway.
- Especially with federated users, it could be a temporary issue. A later lookup could return a user. With offline sessions not being invalidated here lots of users would not need to login again but just wait until the temporary issue is solved.

mhajas replied:
That sounds reasonable @mkrueger92, thanks!

mhajas reviewed (changes requested):
Thank you for the update. I have two comments.
1. See the comment in the code
2. Should we add the user check also to [MapUserSessionProvider](https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java#L91)? The linked line can be changed to something like this:
```java
MapUserSessionAdapter userSessionAdapter = new MapUserSessionAdapter(session, realm, origEntity);
if (userSessionAdapter.getUser() == null) {
    removeUserSession(realm, userSessionAdapter);
    return null;
}
return userSessionAdapter;
```

Commit:
Add user check to MapUserSessionProvider

Only when a user is found create a MapUserSessionAdapter
object. Otherwise, return null.

mhajas reviewed (approved):
Thank you @mkrueger92, the PR LGTM! 

@mposolda @pedroigor Do you want to have another look at this?

mposolda reviewed (approved):
@mkrueger92 Thanks for the update!
@mhajas Thanks for the review!
